### PR TITLE
[Speculative decoding 6/9] Integrate speculative decoding with LLMEngine 

### PR DESCRIPTION
**Pull request by cadedaniel**
This PR integrates the existing speculative decoding components with the LLMEngine. With this PR, one can run speculative decoding with a draft model / target model e2e. Note however that the output is garbage as there remains e2e correctness implementation ([PR 7/9 in the open sourcing plan](https://docs.google.com/document/d/1rE4pr3IdspRw97XbImY4fS9IWYuJJ3HGtL7AdIKGrw8/edit#heading=h.qijw1sdidrer)).

Notes:
* Testing: This PR relies on e2e tests with speculative decoding enabled. We compare that the number of emitted tokens is correct, and that the detokenized string matches what HF transformers produces.
* Testing: this PR includes some unit tests for new detokenization/eos management in LLMEngine
* LLMEngine Refactor: Because it is complicated to combine beam search logic with lookahead scheduling, we introduce a new interface inside the LLMEngine that allows us to process new token ids separately for beam search and lookahead scheduling.
* Spec decode support: Currently the e2e test only passes for TP=1 on GPU. The work to add TP>1 support is trivial, but will be wasted as we need to modify the interfaces for TP>1 anyways. It is left out of this PR.
* This PR modifies the WorkerBase interface slightly, such that all workers return `List[SamplerOutput]` from execute_model instead of just `SamplerOutput`.
* This PR is stacked on top of https://github.com/vllm-project/vllm/pull/3837. A [mirror PR](https://github.com/cadedaniel/vllm-public/pull/3) has been created for easy viewing until that PR is merged.

### CODE REVIEW

**Review by LiuXiaoxuanPKU — COMMENTED** [review]
Thank for the PR! Just some minor comments.

File: vllm/engine/output_processor/multi_step.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+from typing import Callable, Iterable, List
```
Comment (LiuXiaoxuanPKU) [inline vllm/engine/output_processor/multi_step.py:1]: Another question:
From my understanding, multi_step processor is used for target model, which is not a muti_step_worker.
single_step processor is used for the draft model, which is a multi_step_worker. 
Correct me if I'm wrong. If this is the case, can we have a better name for multi_step output processor?

Comment (cadedaniel) [reply vllm/engine/output_processor/multi_step.py:1]: The LLMEngine only interacts with the SpecDecodeWorker, which internally manages workers for the draft and target models. See this diagram from the meetup below:

Screenshot 2024-04-10 at 2 09 30 PM

This means that from the LLMEngine's perspective, there is only one worker type, and only one output processor type at a time.

**Comment by cadedaniel** [comment]
update: seems there is some conflict with https://github.com/vllm-project/vllm/pull/3884 and block manager v2 that only shows up in the spec decode e2e test added in this PR. will investigate.

**Comment by rkooo567** [comment]
^ sorry for the inconvenience @cadedaniel!

**Comment by cadedaniel** [comment]
np glad chunked prefill got in

**Comment by leiwen83** [comment]
Hi @cadedaniel ,

Great job for this feature enabling!
I locally pull your draft patch https://github.com/vllm-project/vllm/pull/3951, and test it for locally with both drafter/verifer use the same model. The correctness of output is confirmed!

However, I find that in https://github.com/vllm-project/vllm/blob/main/vllm/spec_decode/multi_step_worker.py#L347, you append -1 towards the proposal to fill num_lookahead_slots number. Which maybe a bit of waste for the verification stage? Since the proposal may not always num_lookahead_slots length? Is there any improvement space here?

Thx~

**Comment by cadedaniel** [comment]
> Hi @cadedaniel ,
> 
> Great job for this feature enabling! I locally pull your draft patch #3951, and test it with both drafter/verifer use the same model. The correctness of output is confirmed!

Thanks for testing it out.

> However, I find that in https://github.com/vllm-project/vllm/blob/main/vllm/spec_decode/multi_step_worker.py#L347, you append -1 towards the proposal to fill num_lookahead_slots number. Which maybe a bit of waste for the verification stage? Since the proposal may not always num_lookahead_slots length? Is there any improvement space here?
> 
> Thx~

Yes, it would be wasteful if we verified these padding token ids. We avoid this via two means:
* `num_lookahead_slots` is always k during speculation.
* a proposal is either of len `k` or 0 -- we either speculate `k` tokens ahead of do not speculate at all.

This means we never have any `-1` proposal tokens to verify.

In the future, we can support per-sequence proposal lens.

**Review by LiuXiaoxuanPKU — APPROVED** [review]

**Pull request merged**